src/main/java/com/google/devtools/common/options/Converters.java

```diff
@@ -625,4 +632,71 @@ public String getTypeDescription() {
       return "Converts to a CacheBuilderSpec, or null if the input is empty";
     }
   }
+
+  public static class BooleanFormulaConverter implements Converter<Expression> {
+    @Override
+    public Expression convert(String input) throws OptionsParsingException {
+      if (input.isEmpty()) {
+        return Expression.empty();
+      }
+      if (isLegacySyntax(input)) {
+        input = convertLegacySyntaxToBooleanFormula(input);
+      }
+      try {
+        Expression result = Expression.parse(ParserInput.fromLines(input));
+        validate(result);
+        return result;
+      } catch (SyntaxError.Exception e) {
+        throw new OptionsParsingException("Failed to parse expression: " + e.getMessage() + " input: " + input, e);
+      }
+    }
+
+    @Override
+    public String getTypeDescription() {
+      return "A Boolean formula over test tags (e.g. '(release or smoke) and not flaky')";
+    }
+
+    private static boolean isLegacySyntax(String input) {
+      return input.contains(",") || input.contains("-");
+    }
+
+    /**
+     * Converts a string containing tags separated by comma to a boolean expression.
+     */
+    private String convertLegacySyntaxToBooleanFormula(String input) throws OptionsParsingException {
+      return input.replaceAll(" *, *", " or ").replace("-", " not ").trim();
+    }
+
+    /**
+     * Throws an OptionsParsingException when the given boolean formula does not follow the grammar:
+     * expr = IDENT | expr 'or' expr | expr 'and' expr | 'not' expr | '(' expr ')'
+     */
+    private void validate(Expression expression) throws OptionsParsingException {
+      switch (expression.kind()) {
+        case IDENTIFIER:
+          break;
+        case BINARY_OPERATOR:
+          BinaryOperatorExpression boe = (BinaryOperatorExpression) expression;
+          if (boe.getOperator() == TokenKind.OR || boe.getOperator() == TokenKind.AND) {
```

Comment:
Tip: the control structures are slightly easier to read if you reject the error first:

```
   if (op != OR && op != AND) {
            throw ...
    }
    validate(X)
    validate(Y)
    break
```

Ditto for the unary case.